### ISSUE TITLE
compare_images computes RMS incorrectly

### ISSUE THREAD

**Issue by mgiuca-google** [matplotlib/matplotlib#1287]
TL;DR: The image comparison algorithm is broken. I have a [patch branch](https://github.com/mgiuca-google/matplotlib/tree/fix-image-comparison), but it isn't ready because it makes some test cases fail and I would like some input on how to proceed.

The matplotlib.testing.compare.compare_images function, which compares images for the purpose of testing, implements a particularly poor image comparison algorithm which, among other things, is unable to detect movement of pixels between two images. It claims to be calculating the root mean square (implicitly, by the use of the variable name "rms"), but it is not the RMS of the images at all, but rather the RMS of their histograms.

The algorithm currently works this way. To compare expectedImage and actualImage:

```
For each colour component c in (r, g, b):
    Build a histogram of the c values in expectedImage
    Build a histogram of the c values in actualImage
    Find the root mean square of the histogram
Sum the three root mean square values
```

To explain this, let's make the simplifying assumption that the image is greyscale. What this algorithm does is it counts up the total number of pixels with the value 0, the total number with the value 1, up to 255. Then it considers the images to mismatch if these counts are significantly different.

Here are two specific examples of where this approach fails epically. Firstly, imagine you have two images that are extremely similar, but all of the pixels differ by exactly 1 value. For example, [all pixels are 127](https://github.com/mgiuca-google/matplotlib/blob/3a83a246321356ed99f63ac6cec41c4b1829347c/lib/matplotlib/tests/baseline_images/test_compare_images/all127.png) in the first image, [all pixels are 128](https://github.com/mgiuca-google/matplotlib/blob/3a83a246321356ed99f63ac6cec41c4b1829347c/lib/matplotlib/tests/baseline_images/test_compare_images/all128.png) in the second. This algorithm will consider these two images to be maximally different, because the pixels are in totally different histogram buckets.

Secondly, imagine you take [an image](https://github.com/mgiuca-google/matplotlib/blob/3a83a246321356ed99f63ac6cec41c4b1829347c/lib/matplotlib/tests/baseline_images/test_compare_images/cosine_peak-nn-img.png), and then take all of the pixels and [randomly shuffle them around](https://github.com/mgiuca-google/matplotlib/blob/3a83a246321356ed99f63ac6cec41c4b1829347c/lib/matplotlib/tests/baseline_images/test_compare_images/cosine_peak-nn-img-scrambled.png). The algorithm will consider these two images to be exactly identical, because even though the pixels are completely messed up, they generate the exact same histograms. A real-world example of where this will not catch a failure is if a label is, for some reason, printed at the completely wrong coordinates.

The correct approach is to take the RMS of the differences between each corresponding pixel. To do this, use the following algorithm:

```
Subtract the two images from one another, and take the absolute value
Find the root mean square of the delta image
```

This can be optimised using a histogram (take the histogram of the delta image, and then for each bucket, consider there to be that many pixels of that value). I have [prepared a patch](https://github.com/mgiuca-google/matplotlib/tree/fix-image-comparison) which fixes this issue.

Note that my algorithm produces identical results to ImageMagick's `compare -metric rmse` command (once you remove the alpha channels).

The other problem with the current algorithm is that the units are meaningless. The unit of measurement is technically "absolute number of pixels", which means that the magnitude is determined by the image size. For a 10x10 image, the worst RMS possible is 100; for a 640x480 image, it is 307200. This makes it impossible to set a reasonable tolerance value. In the corrected solution, the unit of measurement is "colour value difference", where the worst RMS possible is 255. To cope with this change, I have dropped the division by 10000 and multiplied all of the tolerances by 10000 -- I found that 10 (the default) makes a reasonable tolerance under the new scheme.

Unfortunately, by fixing this algorithm, a number of test cases have started failing. On my machines (two separate machines running Ubuntu 12.04), I am seeing a lot of the text characters rendered slightly off the expected position. This isn't a problem, but it does fail in the new algorithm because the text is in the wrong position.

The vast majority of failing cases are in the mathtext module. These have RMS errors from 10 up to 60. There are three other cases outside of mathtext that now fail: 
- `test_axes.test_single_date`: RMS 12.861
- `test_simplification.test_clipper`: RMS 20.035 -- this is genuinely broken!
- `test_spines.test_spines_axes_positions`: RMS 11.431

I think it would be fine to bump up the tolerance for the first and third one; the `test_clipper` looks to me like it's actually producing incorrect output. **I'd like to know if other users have similar (perhaps exactly the same) RMS errors for these cases, with my patch applied.** If so, it is probably the case that these expected output images are out of date, and we can regenerate them. If not, then perhaps we can bump up the tolerance a bit to allow for differences in the rendering algorithm. (Though I wouldn't want to go as high as 60.)

I have not created a pull request, since the branch is not yet ready to merge (due to the failing cases). However, hopefully we can work towards fixing up those test cases.

Thanks for your time.
